feat: allow customizing browser data location

zcbenz opened the pull request:
#### Description of Change

This PR adds a new `sessionData` path to `app.setPath`, which allows changing the location where browser data like cookies and disk cache are stored.

This is very useful for apps that do not store user information in browser, as they treat browser data as cache and want to store them in cache locations, see also https://github.com/electron/electron/issues/1404 and https://github.com/electron/electron/issues/8124.

Before this PR apps can achieve similar effect by changing the path of `userData`, but we should separate `sessionData` from `userData`, as apps may want to store user information in `userData` while storing browser data in cache location.

This PR is modified from https://github.com/electron/electron/pull/30665.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Add `browserData` to `app.setPath/getPath`.

nornagon commented on shell/app/electron_main_delegate.cc:
```diff
@@ -134,11 +134,18 @@ bool ElectronPathProvider(int key, base::FilePath* result) {
       break;
+    case DIR_BROWSER_DATA: {
+      // By default and for backward, equivalent to DIR_USER_DATA.
+      if (!base::PathService::Get(chrome::DIR_USER_DATA, &cur))
+        return false;
+      create_dir = true;
+      break;
```
```suggestion
      return base::PathService::Get(chrome::DIR_USER_DATA, &cur);
```

I think `chrome::DIR_USER_DATA` already [creates the directory if it doesn't exist](https://source.chromium.org/chromium/chromium/src/+/main:chrome/common/chrome_paths.cc;l=165;drc=99f36136e05dbf96fa947c21feb467f60c357558), so we don't need to repeat the check.

nornagon commented on shell/browser/browser.cc:
```diff
@@ -190,6 +190,9 @@ void Browser::DidFinishLaunching(base::DictionaryValue launch_info) {
   base::FilePath user_data;
   if (base::PathService::Get(chrome::DIR_USER_DATA, &user_data))
     base::CreateDirectoryAndGetError(user_data, nullptr);
+  base::FilePath browser_data;
+  if (base::PathService::Get(DIR_BROWSER_DATA, &browser_data))
+    base::CreateDirectoryAndGetError(browser_data, nullptr);
```
Why is this needed? The directory should be created on demand, right?

zcbenz replied:
If user sets a custom dir, it is not created on demand. There is code ensuring DIR_USER_DATA  to be created here for the same reason.

nornagon commented on docs/api/app.md:
```diff
@@ -637,6 +637,8 @@ Returns `string` - The current application directory.
     * `~/Library/Application Support` on macOS
   * `userData` The directory for storing your app's configuration files, which by
     default it is the `appData` directory appended with your app's name.
+  * `browserData` The directory for storing browsing data, such as localStorage,
+    cookies, and disk cache. By default this points to `userData`.
```
quite confused about this description, because it seems to me that localStorage and cookies should be in the "user data" directory.

zcbenz replied:
Chromium does not provide a way to set separate locations for localStorage and disk cache, so we currently put them all under `userData`. This PR means to provide a way to change the location of localStorage and disk cache, as there are apps willing to put them under cache locations (https://github.com/electron/electron/issues/1404).

In future I would like to add a new `browserCache` path to only set location for things like disk cache, but currently it is not practical as it requires a large patch to Chromium.

nornagon replied:
Then I'm not sure I see the point of this change then? What then is stored in `userData`?

zcbenz replied:
For example in VS Code, user configurations are stored as JSON under `userData`, while there is no user information stored in localStorage or cookies.

Also apart from browser data, things like crash reports and app logs are currently stored under `userData`, the `requestSingleInstance` API also relies on `userData` to exchange data.

For apps don't rely on localStorage or cookies to store user data, it is necessary to separate `browserData` from `userData`.

nornagon replied:
So it looks like this PR would group the following things under the new `browserData` directory:
- dictionaries
- browser prefs (`Local State`). not sure exactly what is stored in here, maybe things like devtools state?
- network service state. also not sure exactly what data is stored here. dns cache perhaps?
- ElectronBrowserContext's path, which is all the user profile data like localStorage, cookies, etc.
- some devtools-related temporary unix sockets

and things that remain in `userData` are:
- crash reports
- some caches?
- anything else?

Is this accurate? This is a weird grouping. I feel like `browserData` is the wrong name for this. At the very least this should be extensively documented, including exactly which data is stored in which directory.

zcbenz replied:
The idea behind the grouping is, users should be able to move all files generated by `BrowserContext` to a different directory. So dictionaries, browser prefs, devtools files, network caches belong to the same group with localStorage and cookies, as they are only generated when a `BrowserContext` is created.

I agree `browserData` is confusing, "browser" means lots of things in Electron. How do you think using `sessionData` as name instead? It should be clear that it is where `Session` code writes files to. If in future we add APIs to specify the path of `Session` to store data, it would point to `sessionData` by default.

  * `sessionData` The directory for storing data generated by `Session`, such as localStorage, cookies, disk cache, downloaded dictionaries, network state, devtools files. By default this points to `userData`.

nornagon replied:
I don't think downloaded dictionaries are specific to a Session?

nornagon replied:
Also, are we sure that no other "session-related" data is stored in userData? e.g. v8 code caches, IndexedDB data, service worker registration, etc. etc.

zcbenz replied:
Dictionaries are downloaded by Chromium automatically via the `ses.setSpellCheckerDictionaryDownloadURL` API, and all dictionary APIs belong to Session. In Chromium's context, the SpellcheckServiceFactory is attached to BrowserContext.

For other data, I can write tests to verify they are in the expected location.

nornagon replied:
I'm somewhat worried about new session-related data being introduced in Chromium with no easy way for us to redirect where it's stored without patching Chromium. It also seems really easy to miss session-related data when implementing this. How can we be sure we caught all the cases?

Conceptually, though, the idea of a `sessionData` path doesn't seem totally unreasonable. It does raise the question from an API perspective of whether it should be possible to set session paths on a per-session basis.

zcbenz commented:
I have renamed `browserData` to `sessionData`, and added tests to verify the locations of service worker, localStorage, etc. are the in the right location.

nornagon commented on docs/api/app.md:
```diff
@@ -637,6 +637,9 @@ Returns `string` - The current application directory.
     * `~/Library/Application Support` on macOS
   * `userData` The directory for storing your app's configuration files, which by
     default it is the `appData` directory appended with your app's name.
```
Let's expand this comment to describe how it differs from `sessionData`, since "user" data is now quite confusing.

jkleinsc reviewed (commented):
API LGTM

nornagon commented on docs/api/app.md:
```diff
@@ -635,8 +635,18 @@ Returns `string` - The current application directory.
+  * `userData` The directory for storing your app's configuration files, which
+    by default is the `appData` directory appended with your app's name. By
+    convention files storing user data should be written to this directory, and
+    it is not recommended to write large files here because some environments
+    may backup this directory to cloud storage.
```
Does Chromium have this problem? I would be surprised if Chrome's caches are uploaded to cloud storage on any platform...

nornagon replied:
Hang on—so Chrome can choose the session data directory per-profile, then! Chromium will put things under `userData/Profile 1`, `userData/Profile 2` and so on. Shouldn't we offer the same capability?

nornagon replied:
Hm, Chrome writes to `User Data/$partitionName`, right? `Default` is a name of a partition.

nornagon replied:
ahhhh, right, got it. Thanks, that makes sense!

nornagon commented on docs/api/app.md:
```diff
@@ -635,8 +635,18 @@ Returns `string` - The current application directory.
+    state, devtools files. By default this points to `userData`. Chromium may
+    write very large disk cache here, so if your app does not rely on browser
+    storage like localStorage or cookies to save user data, it is recommended
+    to set this directory to other locations to avoid polluting the `userData`
```
> other locations

I think it would be a good idea to recommend a particular location here!

jkleinsc commented:
Merging as API review has been approved but is still showing as pending for some reason.
